reasoner/planner/CostEstimator.java

```diff
@@ -0,0 +1,384 @@
+                multivarEstimates.get(concludable).addAll(computeEstimatesFromConcludable(concludable));
+            });
+
+            // TODO: Negateds for the total -cost?
+            // TODO: can we prune the multiVarEstimates stored?
+
+            this.fullAnswerCount = computeCostCover(iterate(multivarEstimates.values()).flatMap(Iterators::iterate).toList(), conjunction.pattern().variables());
+        }
+
+        private List<LocalEstimate> computeEstimatesFromConcludable(Concludable concludable) {
+            if (concludable.isHas()) {
+                return list(estimatesFromHasEdges(concludable.asHas()));
+            } else if (concludable.isRelation()) {
+                return list(estimatesFromRolePlayersAssumingEvenDistribution(concludable.asRelation()));
+            } else {
+                return list();
+            }
+        }
+
+        private long computeInferredUnaryEstimates(ThingVariable generatedVariable) {
+            List<Concludable> relevantConcludables = resolvables()
+                    .filter(resolvable -> isInferrable(resolvable) && resolvable.generating().isPresent() &&
+                        resolvable.generating().get().equals(generatedVariable))
+                    .map(Resolvable::asConcludable)
+                    .toList();
+
+            if (relevantConcludables.isEmpty()) {
+                return 0L;
+            }
+
+            long mostConstrainedAnswer = Long.MAX_VALUE;
+            for (Concludable concludable: relevantConcludables) {
+                Map<Rule, Set<Unifier>> unifiers = logicMgr.applicableRules(concludable);
+                AtomicLong nInferred = new AtomicLong(0L);
+                long answersGeneratedByThisResolvable = iterate(unifiers.keySet()).map(rule -> {
+                    Optional<Unifier> firstUnifier = unifiers.get(rule).stream().findFirst();
+                    // Increment the estimates for variables generated by the rule
+                    assert rule.conclusion().generating().isPresent() && firstUnifier.isPresent();
+
+                    if (rule.conclusion().isRelation()) { // Is a relation
+                        Set<Variable> answerVariables = iterate(rule.conclusion().pattern().variables())
+                                .filter(v -> rule.condition().pattern().variables().contains(v)).toSet();
+                        return costEstimator.estimateAnswers(rule.condition().conjunction(), answerVariables);
+                    } else if (rule.conclusion().isExplicitHas()) {
+                        return 1L;
+                    }  else {
+                        assert rule.conclusion().isVariableHas();
+                        return 0L;
+                    }
+                }).reduce(0L, Long::sum);
+
+                mostConstrainedAnswer = Math.min(mostConstrainedAnswer, answersGeneratedByThisResolvable);
+            }
+            assert mostConstrainedAnswer != Long.MAX_VALUE;
+            return mostConstrainedAnswer;
+        }
```

Comment:
For attribute types:, adds one to the unary estimate for each ExplicitHas rule concluding the type (Even if it already occurs in the data)
For relation-types: Sums across each rule concluding a matching type,  the estimated answers to the rule (i.e., to the condition of the rule projected to the variables in the head of the rule)